packages/core/src/use-params.ts

```diff
@@ -90,38 +94,13 @@ export function useParams(returnType?: "string" | "number" | "array") {
 export function useParam(key: string): undefined | string | string[]
 export function useParam(key: string, returnType: "string"): string | undefined
 export function useParam(key: string, returnType: "number"): number | undefined
-export function useParam(key: string, returnType: "array"): Array<string | undefined>
+export function useParam(key: string, returnType: "array"): string[] | undefined
 export function useParam(
   key: string,
-  returnType?: "string" | "number" | "array",
-): undefined | number | string | Array<string | undefined> {
-  const params = useParams()
-  const rawValue = params[key]
-
-  if (returnType === "number") {
-    // Special case because Number("") === 0
-    if (rawValue === "") {
-      return NaN
-    }
-    return Number(rawValue)
-  }
-
-  if (returnType === "string") {
-    if (typeof rawValue !== "string") {
-      return ""
-    }
-    return rawValue
-  }
-
-  if (returnType === "array") {
-    if (typeof rawValue === "undefined") {
-      return []
-    }
-    if (!Array.isArray(rawValue)) {
-      return [rawValue]
-    }
-    return rawValue
-  }
+  returnType?: ReturnTypes,
+): undefined | number | string | string[] {
+  const params = useParams(returnType as any) // unsure why i need to typecast here
```

Comment:
❓ ❗ 

Unsure why I need to typecast as any here and why typescript doesn't get it. Haven't used method overloading in TS much before so there is prob some quirk.

Reply:
When changing to

```suggestion
  const params = useParams(returnType)
```

I get the following error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '"string" | "number" | "array" | undefined' is not assignable to parameter of type 'undefined'.
      Type '"string"' is not assignable to type 'undefined'.ts(2769)
use-params.ts(42, 17): The last overload is declared here.
use-params.ts(47, 17): The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
```

Reply:
It looks like what you want is a generic function. You could probably replace lines 42-45 with just this:

```
export function useParams<T>(returnType: T): Dict<T>
```

Also, I think your definition on line 41 already covers the definition on line 42.

Reply:
Thanks @dmzza. I've played around a bit with generics but couldn't get it to work either. `Dict<T>` won't work as it's string literals and not actually the type that's passed in.


If you - or anyone else watching - wanna clone it and play around, here's a little shortcut:

```sh
git clone git@github.com:KATT/blitz.git katt-blitz
cd katt-blitz/packages/core
git checkout feature/return-strict-or-undefined
yarn
yarn test:watch
```

Reply:
just add like this in `useParams` and all will be good!

```typescript
export function useParams(): Dict<string | string[]>
export function useParams(returnType?: ReturnTypes): Dict<string | string[]>
export function useParams(returnType: "string"): Dict<string>
export function useParams(returnType: "number"): Dict<number>
export function useParams(returnType: "array"): Dict<string[]>
```

Reply:
Amazing! Thank you @ryandrews 🙏